llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -6855,6 +6855,50 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
   if (Value *V = simplifyICmpInst(I.getPredicate(), Op0, Op1, Q))
     return replaceInstUsesWith(I, V);
 
+  {
+    Value *X;
+    const APInt *C1, *C2;
+    ICmpInst::Predicate Pred;
+    if ((match(&I, m_ICmp(Pred, m_SRem(m_Value(X), m_NonNegative(C1)),
+                          m_APInt(C2))) &&
+         ((Pred == ICmpInst::ICMP_SLT && *C2 == *C1 - 1) ||
+          (Pred == ICmpInst::ICMP_SGT && *C2 == *C1 - 2))) ||
+        (match(&I,
+               m_ICmp(Pred, m_SRem(m_Value(X), m_Negative(C1)), m_APInt(C2))) &&
+         ((Pred == ICmpInst::ICMP_SGT && *C2 == *C1 + 1) ||
+          (Pred == ICmpInst::ICMP_SLT && *C2 == *C1 + 2)))) {
+      // icmp slt (X s% C), (C - 1) --> icmp ne (X s% C), (C - 1), if C >= 0
+      // icmp sgt (X s% C), (C - 2) --> icmp eq (X s% C), (C - 1), if C >= 0
+      // icmp sgt (X s% C), (C + 1) --> icmp ne (X s% C), (C + 1), if C < 0
+      // icmp slt (X s% C), (C + 2) --> icmp eq (X s% C), (C + 1), if C < 0
+      return CmpInst::Create(Instruction::ICmp,
+                             ((Pred == ICmpInst::ICMP_SLT && *C2 == *C1 - 1) ||
+                              (Pred == ICmpInst::ICMP_SGT && *C2 == *C1 + 1))
+                                 ? ICmpInst::ICMP_NE
+                                 : ICmpInst::ICMP_EQ,
+                             Op0,
+                             ConstantInt::get(Op1->getType(), C1->isNegative()
+                                                                  ? *C1 + 1
+                                                                  : *C1 - 1));
+    }
+
+    if (match(&I, m_ICmp(Pred, m_URem(m_Value(X), m_NonNegative(C1)),
```

Comment:
Please remove the support for signed icmps and the nneg constraint.
```
icmp ult (X u% C), (C - 1) --> icmp ne (X u% C), (C - 1)
icmp ugt (X u% C), (C - 2) --> icmp eq (X u% C), (C - 1), if C > 1
```
Alive2: https://alive2.llvm.org/ce/z/LMvAC6